### ISSUE TITLE
[ONNX] Cannot turn off Conv+BN fusion in eval mode

### ISSUE THREAD

**Issue by shinh** [pytorch/pytorch#49226]
## 🚀 Feature
We would like an option in `torch.onnx.export` which turns off [Conv+BN fusion](https://github.com/pytorch/pytorch/pull/40547).

## Motivation

We are using ONNX for training. When our trainer starts, it exports a pytorch model in both eval mode and training mode and then compile them into native code. Since the compilation takes long time, we want to reuse the compiled code, replacing only parameters. However, this Conv+BN makes it hard to find corresponding pairs of parameters in eval model and training model.

## Pitch

Like `do_constant_folding`, it would be nice to have an option to turn off the optimization. How about adding `do_eval_peephole` for example?

## Alternatives

If `torch.onnx` team is planning to add more optimization passes, I think it will be more useful to allow users to specify optimization options one by one. E.g.,

```
torch.onnx.export(model, args, 'foo.onnx', optimizations=[torch.onnx.constant_folding, torch.onnx.conv_bn_fusion])
```

## Additional context

This is an unrelated topic, but personally, I don't quite understand why torch.onnx is implementing optimization passes. I think many ONNX backend already has optimizations like Conv+BN fusion. Also, ONNX-to-ONNX optimization can be done in a separate library like https://github.com/onnx/optimizer .

cc @houseroad @spandantiwari @lara-hdr @BowenBao @neginraoof

**Comment by jspisak** [comment]
cc @spandantiwari

**Comment by jspisak** [comment]
cc @SplitInfinity .. :)

**Comment by spandantiwari** [comment]
If you export the model in training mode, `torch.onnx.export(..., training=TrainingMode.TRAINING,...)`, you would see that none of the optimizations are applied and you'd even find a initializers in the ONNX graph corresponding to the PyTorch model parameters (in almost all cases).

By design, we try not to do any mutations in the initializers/parameters in the ONNX graph in the training mode export (as above). But for eval mode, we go the opposite way, and choose do some optimizations/mutations as necessary, such as constant folding and conv+BN fusion. In the past we have had feedback on providing more optimized graphs for eval out of the box.

**Comment by shinh** [comment]
OK, it sounds like `TrainingMode.EVAL` is inference mode designed for models with all params fixed and we should use `TrainingMode.TRAINING` rather than `TrainingMode.EVAL` for eval during training.

Another problem is `TrainingMode` is coupled with the mode of dropout/BN. If we unconditionally use `TrainingMode.TRAINING`, making it possible to use fixed BN in `TrainingMode.TRAINING` will be more important for us. We can discuss it in https://github.com/pytorch/pytorch/issues/49229

**Comment by spandantiwari** [comment]
OK. So let's track the other issue in 49229.

**Comment by SystemErrorWang** [comment]
> If you export the model in training mode, `torch.onnx.export(..., training=TrainingMode.TRAINING,...)`, you would see that none of the optimizations are applied and you'd even find all initializers in the ONNX graph corresponding to the PyTorch model parameters (in almost all cases).
> 
> By design, we try not to do any mutations in the initializers/parameters in the ONNX graph in the training mode export (as above). But for eval mode, we go the opposite way, and choose do some optimizations/mutations as necessary, such as constant folding and conv+BN fusion. In the past we have had feedback on providing more optimized graphs for eval out of the box.

Is there any method to turn the BatchNorm node in training mode to inference mode in an onnx file? Or if there is any method to export BatchNorm nodes with dynamic input in inference mode? (refer to adain implementaiton in this repo: https://github.com/NVlabs/MUNIT/blob/master/networks.py)

**Comment by osayes** [comment]
A question:
Why the `moving_mean` and `moving _var` can't be normal parameters/weights? The optimization doesn't depend on their specific value, and if I want to update the model, I will most likely update them too.

**Comment by francis2tm** [comment]
@fxmarty  exatactly. I'm having exactly that problem, I don't want to disable optimization, I just want no fuse of Conv and BatchNorm.
Did you manage to find a solution?

**Comment by fxmarty** [comment]
@francis2tm I don't recall solving the issue.  What I recall is having a case where I needed to fold manually because the ONNX export wouldn't: https://github.com/microsoft/onnxruntime/issues/12522#issuecomment-1209618944
